src/snowflake/snowpark/_internal/server_connection.py

```diff
@@ -655,7 +665,24 @@ def _fix_pandas_df_integer(
             and column_metadata.scale == 0
             and not str(pandas_dtype).startswith("int")
         ):
-            pd_df[pandas_col_name] = pandas.to_numeric(
-                pd_df[pandas_col_name], downcast="integer"
-            )
+            # pandas.to_numeric raises "RuntimeWarning: invalid value encountered in cast"
+            # when it tried to downcast and loses precision. In this case, we try to convert to
+            # int64 using astype() and fallback to to_numeric if we were unsuccessful.
+            with warnings.catch_warnings(record=True) as warning_list:
+                pd_col_with_numeric_downcast = pandas.to_numeric(
+                    pd_df[pandas_col_name], downcast="integer"
+                )
+
+            if (
+                len(warning_list) == 1
+                and warning_list[0].message
+                == "RuntimeWarning('invalid value encountered in cast')"
+            ):
```

Comment:
Can we just use the `try-catch` below and remove this `if-else`? It seems like we want to use `astype` whenever possible

Reply:
I had this approach earlier but @sfc-gh-yixie pointed out that it can be a breaking change. It causes failure is some test cases where `pd.to_numeric()` with fit data into a pandas column of `int8` if the data is small enough to fit into it, whereas if we use `astype` whenever possible, we will change our old behavior.